File: leetcode/helpers/bTreeGenerator.js
function TreeNode(val, left, right) {
  this.val = (val === undefined ? 0 : val)
  this.left = (left === undefined ? null : left)
  this.right = (right === undefined ? null : right)
}

const generateBTree = (arr) => {
  if (!arr) return null;

  const valQueue = [];
  const nodeQueue = [];

  for (const val of arr) {
    valQueue.push(val);
  }

  const node = new TreeNode(valQueue.shift());
  nodeQueue.push(node);

  while (valQueue.length > 0) {
    const currentNode = nodeQueue.shift();

    let currentVal = valQueue.shift();
    if (currentVal) {
      left = new TreeNode(currentVal);
      currentNode.left = left;
      nodeQueue.push(left);
    }

    currentVal = valQueue.shift();
    if (currentVal) {
      right = new TreeNode(currentVal);
      currentNode.right = right;
      nodeQueue.push(right);
    }
  }

  return node;
}

// const root = generateBTree([1, 2, 2, 3, null, null, 3, 4, null, null, 4]);
// console.log(root.val);
// console.log(root.left.val);
// console.log(root.right.val);
// console.log(root.left.left.val);

module.exports = generateBTree;

File: leetcode/helpers/bTreeGenerator.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def generateBTree(data):
    if data == None or len(data) == 0:
        return None

    nodeQueue = []
    intQueue = []

    for i in range(1, len(data)):
        intQueue.append(data[i])

    node = TreeNode(data[0])
    nodeQueue.append(node)

    while intQueue:
        leftVal = intQueue.pop(0)
        rightVal = intQueue.pop(0)

        currentNode = nodeQueue.pop(0)

        left = TreeNode(leftVal)
        currentNode.left = left
        nodeQueue.append(left)

        right = TreeNode(rightVal)
        currentNode.right = right
        nodeQueue.append(right)

    return node


# tree = creatBTree([1, 2, 2, 3, None, 1, 3, 4, 8, 1, 4])

# print(tree.val)
# print(tree.left.right.val)
# print(tree.right.val)
